src/marzullo.zig

```diff
@@ -0,0 +1,306 @@
+            if (previous) |p| {
+                assert(p.offset <= tuple.offset);
+                if (p.offset == tuple.offset) {
+                    if (p.bound != tuple.bound) {
+                        assert(p.bound == .lower and tuple.bound == .upper);
+                    } else {
+                        assert(p.source < tuple.source);
+                    }
+                }
+            }
+            previous = tuple;
+
+            // Update the current number of overlapping intervals:
+            count -= switch (tuple.bound) {
+                .lower => @as(i64, -1),
+                .upper => @as(i64, 1),
+            };
+            // The last upper bound tuple will have a count of one less than the lower bound.
+            // Therefore, we should never see count >= best for the last tuple:
+            if (count > best) {
+                best = count;
+                interval.lower_bound = tuple.offset;
+                interval.upper_bound = tuples[i + 1].offset;
+            } else if (count == best and tuples[i + 1].bound == .upper) {
+                // This is a tie for best overlap. Both intervals have the same number of sources.
+                // We want to choose the smaller of the two intervals:
+                const alternative = tuples[i + 1].offset - tuple.offset;
+                if (alternative < interval.upper_bound - interval.lower_bound) {
+                    interval.lower_bound = tuple.offset;
+                    interval.upper_bound = tuples[i + 1].offset;
+                }
+            }
+        }
+        assert(previous.?.bound == .upper);
+
+        // The number of false sources (ones which do not overlap the optimal interval) is the
+        // number of sources minus the value of `best`:
+        assert(best >= 0);
+        assert(best <= sources);
+        interval.sources_true = @intCast(u8, best);
+        interval.sources_false = @intCast(u8, sources - @intCast(u8, best));
+        assert(interval.sources_true + interval.sources_false == sources);
+
+        return interval;
+    }
+
+    /// Sorts the list of tuples by clock offset. If two tuples with the same offset but opposite
+    /// bounds exist, indicating that one interval ends just as another begins, then a method of
+    /// deciding which comes first is necessary. Such an occurrence can be considered an overlap
+    /// with no duration, which can be found by the algorithm by sorting the lower bound before the
+    /// upper bound. Alternatively, if such pathological overlaps are considered objectionable then
+    /// they can be avoided by sorting the upper bound before the lower bound.
+    fn less_than(context: void, a: Tuple, b: Tuple) bool {
+        if (a.offset < b.offset) return true;
+        if (b.offset < a.offset) return false;
+        if (a.bound == .lower and b.bound == .upper) return true;
+        if (b.bound == .lower and a.bound == .upper) return false;
+        if (a.source < b.source) return true;
```

Comment:
what does `source` represent if we are sorting on it?

Reply:
Given two identical offset samples from two clock sources (i.e. two replicas), break the tie deterministically to sort replica 0 ahead of replica 1 etc. It's just to make sure that the sort is fully specified and deterministic. I will comment that here too.